### ISSUE TITLE
Graal(.js) learning resources

### ISSUE THREAD

**Issue by DGolubets** [oracle/graal#331]
Hi guys, the whole project looks so promising! But it's very hard to find any learning material. Do you have any links to tutorials or examples for application developer? Or is there a Gitter channel or user group?

I'm working on an isomorphic web frontend with React.js. I've made a prototype with Nashorn, but I'm not quite satisfied with the performance: it takes too long to warm up to reach Node.js rendering times. So naturally I would like to try Graal.js and see how it competes. However, I don't know where to start.

Basically I need to figure out 2 steps:
1) start my java app with Graal instead of hotspot (correct me, but I think I won't be able to integrate Graal.js if I use Hotspot?).
2) from java code create Graal.js script engine (so I'm looking for something similar to Nashorn API)

Though unlikely, but if by a chance you have a link or example of integrating it with scala\sbt that would be awesome.

**Comment by chumer** [comment]
Ad 1) You should use the GraalVM[1] binary it comes with Graal as the default second tier compiler.
Ad 2) With GraalVM we ship a bridge for the javax.scripting API. But we recommend to use our polyglot API instead[1].

[1] http://www.oracle.com/technetwork/oracle-labs/program-languages/downloads/download-2301612.html
[2]  Primer: https://github.com/oracle/graal/blob/master/sdk/docs/PolyglotEmbedding.md. Javadoc is here: http://www.graalvm.org/sdk/javadoc/org/graalvm/polyglot/package-summary.html

More extensive documentation is coming very soon.

**Comment by DGolubets** [comment]
Thanks @chumer !

**Comment by chumer** [comment]
Ah missed your last question. You don't need to do anything for scala\sbt integration besides using GraalVM as your JDK to run. The polyglot API is available there as well without any changes (without additional classpath entries).

**Comment by DGolubets** [comment]
Sounds easy :)

Is there any plan to release GraalVM for Windows? I noticed that there are no binaries for it.

**Comment by chumer** [comment]
There are plans, but I cannot provide you with an ETA just yet.
Graal.JS itself works on Windows (but slow(without Graal optimizations)).

**Comment by chumer** [comment]
Let us know about the warmup numbers. We also have some room for improvement there. We are also working on a feature that lets you share/cache the warmup(parsing, compilation) across multiple polyglot contexts (ScriptEngine in javax.scripting terms). Similar to what Rhino can do.

This feature will likely be done in the one after the next release. (~1-2 months)

**Comment by DGolubets** [comment]
Here are results of my (completely non-scientific) benchmark:

5 iterations:
- Node: 2ms
- Nashorn: 36ms
- Graal: 23ms

100 iterations:
- Nashorn: 28ms
- Graal: 15ms

1000 iterations:
- Nashorn: 17ms
- Graal: 7ms

10000 iterations:
- Nashorn: 11ms
- Graal: 2ms

So, it looks like Graal.js is definitely an improvement over Nashorn in performance. But it still takes too long to warm up to get to Node.js results.

**Comment by DGolubets** [comment]
Well, basically these numbers are average over N iterations (excluding very first one). I guess I should exclude warmup cycles from measurement..

So, second approach is N warmup iterations and then average time over 10 test iterations:
```scala

  def benchmark(times: Int, warmUp: Int)(code: => Unit): Duration = {
    for (i <- 1 to warmUp) {
      code
    }

    val t1 = System.nanoTime
    for (i <- 1 to times) {
      code
    }
    val t2 = System.nanoTime
    Duration.fromNanos((t2 - t1) / times)
  }
  val moduleText = io.Source.fromFile("path to some JS module built with WebPack").mkString
  context.eval("js", moduleText)
  val t = benchmark(N, M) {
    context.eval("js", "ecommerce.renderToString({ title: \"abc\"})")
  }
```

5 iterations:
- Nashorn: 42ms
- Graal: 24ms

100 iterations:
- Nashorn: 24ms
- Graal: 12ms

1000 iterations:

- Nashorn: 13ms
- Graal: 3ms

This looks better :)

That's all on JVM and I think these warmup iterations should be enough for both compiler and interpreter.

**Comment by chumer** [comment]
Thanks a lot for sharing! 

As I said expect this warmup curve to be come steeper in the next months.

**Comment by DGolubets** [comment]
Hi @wirthi ,
I will prepare some standalone benchmark code later. If it could help the team to optimize Graal.js more, that would be great!
Thanks guys!

**Comment by wirthi** [comment]
I am closing this issue. If you have any benchmark to share, please feel free to open a new issue for it (easiest to track for the team if you do so on https://github.com/graalvm/graaljs).
